src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ImmutableArrayExtensions.cs

```diff
@@ -11,6 +11,10 @@ internal static partial class ImmutableArrayExtensions
     {
         public static ImmutableArray<T> ToImmutableArray<T>(this HashSet<T> set)
         {
+            // [.. set] currently allocates, even for the empty case.  Workaround that until that is solved by the compiler.
+            if (set.Count == 0)
```

Comment:
fyi @RikkiGibson @DoctorKrolic

Reply:
It looks like we're allocating with `new T[set.Count]`. Does that actually result in a heap allocation if `set.Count == 0`?

Reply:
yes.  `new T[0]` is a fresh array allocation.

Reply:
I would be ok with codegen changing to call an appropriate `ToImmutableArray` extension here.

I'm less sure about branching on the count. Maybe it's a good idea? Do we have a runtime helper which will alloc an array for non-empty and otherwise give the "canonical" empty array?

I do also find it a little surprising that we are doing GetEnumerator/MoveNext here instead of ICollection.CopyTo. Maybe we never implemented the support for calling ICollection.CopyTo, we were just relying on the impl of List.AddRange using it.

There are quite a few [backlog issues](https://github.com/dotnet/roslyn/issues?q=is%3Aopen+is%3Aissue+label%3A%22Code+Gen+Quality%22+label%3A%22New+Feature+-+Collection+Expressions%22) in this category and a broad set of scenarios. It will take a significant amount of time for things to reach a very high level of quality.

I think it's at the level of complexity that a design doc is likely needed, which breaks the scenarios down into categories, and explains how exactly we want to optimize in all those scenarios and why. There have been so many times where I spent time searching for a one-off comment where I laid out how we should do some optimization in this space. That stuff needs to be centralized :)

Reply:
> I would be ok with codegen changing to call an appropriate ToImmutableArray extension here.

Or perhaps a helper we emit ourselves if not present.

> I'm less sure about branching on the count. Maybe it's a good idea? 

It def seems like a good idea to me.  It avoids an alloc, which is going to be more of a hit on the system than just a branch.  

> I do also find it a little surprising that we are doing GetEnumerator/MoveNext here instead of ICollection.CopyTo

This i care less about.  but it would be good to get some numbers to see if it's better or not.

Reply:
Given that `HashSet<T>` has an implicit reference conversion to `IEnumerable<T>` and it implements `ICollection<T>`, this falls under [the following issue](https://github.com/dotnet/roslyn/issues/71296). Therefore the expected codegen after the issue is resolved to be `set.ToImmutableArray()` (or for this particular case it is better to call `ImmutableArray.CreateRange(set)`).

> > I'm less sure about branching on the count. Maybe it's a good idea?
>
> It def seems like a good idea to me. It avoids an alloc, which is going to be more of a hit on the system than just a branch.

I don't agree. Roslyn's goal is to maintain balance between emitting perfectly optimized IL and the size of this IL. In many cases it is better to leave some optimizations to the JIT or runtime libraries and I believe this is exactly that case. Why do we need to emit a count check for _every_ spread like this if this check can be done inside `ImmutableArray.CreateRange()` method? It would be then up to the JIT to maybe inline this call, see, that `HashSet<T>` always implements `ICollection<T>`, therefore it can eliminate the code path for non-`ICollection<T>` cases.

> Or perhaps a helper we emit ourselves if not present.

On modern runtimes we can expect that `ToImmutableArray`/`CreateRange` are always present. For older runtimes like .NET Framework we can state the following:
1. Collection expressions are a C#12+ feature and C#12+ versions aren't officially supported on those old runtimes, therefore if compiler codegen isn't perfect there we can always say "Well, you are doing something "unsupported", all responsibility is on you".
2. .NET Framework doesn't have `System.Collections.Immutable` built-in. If user wants to use these types there, they need to bring them via a NuGet package. And since that NuGet package is updated with every .NET release, its latest version definitely has desired helper methods.

So in the end I don't think maintaining such complicated fallback scenario is worth implementation and maintainability efforts here.

Reply:
> I don't agree

My point was that it was important to emit code that allows the alloc to be avoided.  So we should not emit a new array allocation in all cases.  How that happens is less relevant to me.